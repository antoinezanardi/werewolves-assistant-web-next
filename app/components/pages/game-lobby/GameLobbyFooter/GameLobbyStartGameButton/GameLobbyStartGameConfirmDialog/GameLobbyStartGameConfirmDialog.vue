<template>
  <PrimeVueConfirmDialog
    id="game-lobby-start-game-confirm-dialog"
    ref="gameLobbyStartGameConfirmDialog"
    :pt="{ 'root': 'w-1/3' }"
  >
    <template #container="{ acceptCallback, rejectCallback }">
      <GameLobbyStartGameConfirmDialogContainer
        id="game-lobby-start-game-confirm-dialog-container"
        :current-confirm-step="currentConfirmStep"
        :reject-callback="rejectCallback"
        @confirm-start-game="acceptCallback"
        @confirm-step="confirmStep"
        @reject-players-position-step="rejectPlayersPositionStep"
        @reject-start-game="rejectCallback"
      />
    </template>
  </PrimeVueConfirmDialog>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";
import { useConfirm } from "primevue/useconfirm";
import type { ComponentPublicInstance } from "vue";
import type { GameLobbyStartGameConfirmDialogEmits, GameLobbyStartGameConfirmDialogExposed, GameLobbyStartGameConfirmDialogStep } from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/game-lobby-start-game-confirm-dialog.types";
import GameLobbyStartGameConfirmDialogContainer from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/GameLobbyStartGameConfirmDialogContainer.vue";
import { useArrays } from "~/composables/misc/useArrays";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const emit = defineEmits<GameLobbyStartGameConfirmDialogEmits>();

const gameLobbyStartGameConfirmDialog = ref<ComponentPublicInstance | null>(null);

const confirmStepIndex = ref<number>(0);

const createGameDtoStore = useCreateGameDtoStore();
const { doesCreateGameDtoContainPositionDependantRoles } = storeToRefs(createGameDtoStore);

const { require: confirmRequire } = useConfirm();

const { insertIf } = useArrays();

const confirmSteps = computed<GameLobbyStartGameConfirmDialogStep[]>(() => [
  ...insertIf<GameLobbyStartGameConfirmDialogStep>(doesCreateGameDtoContainPositionDependantRoles.value, "players-positioned"),
  "players-ready",
]);

const currentConfirmStep = computed<GameLobbyStartGameConfirmDialogStep>(() => confirmSteps.value[confirmStepIndex.value]);

function open(): void {
  if (!gameLobbyStartGameConfirmDialog.value) {
    throw createError("Game Lobby Start Game Confirm Dialog is not defined");
  }
  confirmStepIndex.value = 0;
  confirmRequire({
    target: gameLobbyStartGameConfirmDialog.value.$el as HTMLElement,
    position: "bottom",
    accept: confirmStartGame,
  });
}

function confirmStep(): void {
  confirmStepIndex.value++;
}

function confirmStartGame(): void {
  emit("confirmStartGame");
}

function rejectPlayersPositionStep(rejectCallback: () => void): void {
  rejectCallback();
  emit("rejectPlayersPositionStep");
}

defineExpose<GameLobbyStartGameConfirmDialogExposed>({
  open,
});
</script>