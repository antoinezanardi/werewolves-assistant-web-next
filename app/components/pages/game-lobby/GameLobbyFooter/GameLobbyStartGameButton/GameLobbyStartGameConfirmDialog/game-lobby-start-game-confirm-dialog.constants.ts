const GAME_LOBBY_START_GAME_CONFIRM_DIALOG_STEPS = [
  "players-positioned",
  "players-ready",
] as const;

export { GAME_LOBBY_START_GAME_CONFIRM_DIALOG_STEPS };