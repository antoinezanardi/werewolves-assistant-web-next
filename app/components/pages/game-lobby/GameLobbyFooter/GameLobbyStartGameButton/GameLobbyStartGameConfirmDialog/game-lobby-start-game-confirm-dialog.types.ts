import type { TupleToUnion } from "type-fest";

import type { GAME_LOBBY_START_GAME_CONFIRM_DIALOG_STEPS } from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/game-lobby-start-game-confirm-dialog.constants";

type GameLobbyStartGameConfirmDialogEmits = {
  confirmStartGame: [];
  rejectPlayersPositionStep: [];
  rejectThiefAdditionalCardsPlacedStep: [];
  rejectActorAdditionalCardsPlacedStep: [];
  rejectGameOptionsChangedStep: [];
};

type GameLobbyStartGameConfirmDialogExposed = {
  open: () => void;
};

type GameLobbyStartGameConfirmDialogStep = TupleToUnion<typeof GAME_LOBBY_START_GAME_CONFIRM_DIALOG_STEPS>;

export type {
  GameLobbyStartGameConfirmDialogEmits,
  GameLobbyStartGameConfirmDialogExposed,
  GameLobbyStartGameConfirmDialogStep,
};