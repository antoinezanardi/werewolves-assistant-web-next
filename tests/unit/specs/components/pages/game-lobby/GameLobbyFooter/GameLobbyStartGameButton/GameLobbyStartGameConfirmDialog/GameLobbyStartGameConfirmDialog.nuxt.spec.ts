import { createTestingPinia } from "@pinia/testing";
import { createFakeCreateGamePlayerDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game/create-game-player/create-game-player.dto.factory";
import { createFakeCreateGameDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game/create-game.dto.factory";
import type { mount } from "@vue/test-utils";

import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import GameLobbyStartGameConfirmDialog from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialog.vue";
import { StoreIds } from "~/stores/enums/store.enum";

describe("Game Lobby Start Game Confirm Dialog Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameLobbyStartGameConfirmDialog>>;
  const defaultCreateGameDto = createFakeCreateGameDto({
    players: [
      createFakeCreateGamePlayerDto(),
      createFakeCreateGamePlayerDto(),
      createFakeCreateGamePlayerDto(),
      createFakeCreateGamePlayerDto(),
    ],
  });
  const testingPinia = { initialState: { [StoreIds.CREATE_GAME_DTO]: { createGameDto: defaultCreateGameDto } } };

  async function mountGameLobbyStartGameConfirmDialogComponent(options: ComponentMountingOptions<typeof GameLobbyStartGameConfirmDialog> = {}):
  Promise<ReturnType<typeof mount<typeof GameLobbyStartGameConfirmDialog>>> {
    return mountSuspendedComponent(GameLobbyStartGameConfirmDialog, {
      shallow: false,
      global: {
        plugins: [createTestingPinia(testingPinia)],
      },
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameLobbyStartGameConfirmDialogComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });
});