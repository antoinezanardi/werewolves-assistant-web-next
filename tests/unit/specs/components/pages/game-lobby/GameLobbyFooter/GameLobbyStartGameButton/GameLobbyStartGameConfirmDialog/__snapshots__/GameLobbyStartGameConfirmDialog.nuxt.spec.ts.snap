// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Start Game Confirm Dialog Component > should match snapshot when rendered. 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
